src/Nonlinear/types.jl

```diff
@@ -0,0 +1,91 @@
+@enum(
+    NodeType,
+    # Index into multivariate operators
+    NODE_CALL_MULTIVARIATE,
+    # Index into univariate operators
+    NODE_CALL_UNIVARIATE,
+    # Index into logic operators
+    NODE_LOGIC,
+    # Index into comparison operators
+    NODE_COMPARISON,
+    # Index is the value of `MOI.VariableIndex`. This is from the original
+    # model, and is not consecutive.
+    NODE_MOI_VARIABLE,
+    # Index of the internal, consecutive, and ordered `MOI.VariableIndex`.
+    NODE_VARIABLE,
+    # Index is into the list of constants
+    NODE_VALUE,
+    # Index is into the list of parameters
+    NODE_PARAMETER,
+    # Index is into the list of subexpressions
+    NODE_SUBEXPRESSION,
+)
+
+struct Node
+    type::NodeType
+    index::Int
+    parent::Int
+end
+
+struct NonlinearExpression
+    nodes::Vector{Node}
+    values::Vector{Float64}
+    NonlinearExpression() = new(Node[], Float64[])
+end
+
+function Base.:(==)(x::NonlinearExpression, y::NonlinearExpression)
+    return x.nodes == y.nodes && x.values == y.values
+end
+
+struct NonlinearConstraint
+    expression::NonlinearExpression
+    set::Union{
```

Comment:
This could be a `MOI.Utilities.StructOfConstraint` with the 4 sets and then a concrete `MOI.Utilities.VectorOfConstraint{NonlinearExpression,S}` in each field of the `MOI.Utilities.StructOfConstraint`

Reply:
This can be a performance optimization at a later date. Let's not get too complicated too soon.